src/ir/flat.h

```diff
@@ -77,8 +80,10 @@ inline void verifyFlatness(Function* func) {
       if (isControlFlowStructure(curr)) {
         verify(!curr->type.isConcrete(),
                "control flow structures must not flow values");
-      } else if (curr->is<LocalSet>()) {
-        verify(!curr->type.isConcrete(), "tees are not allowed, only sets");
+      } else if (auto* set = curr->dynCast<LocalSet>()) {
+        verify(!set->type.isConcrete(), "tees are not allowed, only sets");
+        verify(!isControlFlowStructure(set->value),
+               "set values cannot be control flow");
```

Comment:
Should this be handled in `LocalSet` separately? I mean, does this not work?
```cpp
if (isControlFlowStructure(curr)) {
  verify(!curr->type.isConcrete(),
         "control flow structures must not flow values");
} else {
  if (auto* set = curr->dynCast<LocalSet>()) {
    verify(!set->type.isConcrete(), "tees are not allowed, only sets");
  }
  for (auto* child : ChildIterator(curr)) {
    verify(Properties::isConstantExpression(child) ||                      
             child->is<LocalGet>() || child->is<Unreachable>(),            
           "instructions must only have constant expressions, local.get, " 
           "or unreachable as children");
  }
}
```

Reply:
And I also think the `LocalSet` part checks not on `isConcrete()` but on if the expression is `none`.. Because `LocalSet`'s argument can be `unreachable`, so if `ReFinalize` runs between Flatten and this verification, this can fail. (This does not happen now, but I guess if this property is not satisfied after refinalization, that's not very desirable?)

Reply:
About the first question, set child checking must be separate, as a set's children can be more things (like a call - basically anything that has a non-optional return value; optional return values, that can be worked around, like block return values, are disallowed).

About the last comment, maybe it could be clearer as `!set->isTee() || set->type == unreachable`? Or do you think that isn't right either?

Reply:
> About the first question, set child checking must be separate, as a set's children can be more things (like a call - basically anything that has a non-optional return value; optional return values, that can be worked around, like block return values, are disallowed).

Ah, you're right.
 
> About the last comment, maybe it could be clearer as `!set->isTee() || set->type == unreachable`? Or do you think that isn't right either?

No it looks OK and clearer.